source code info for interpreted options; fix source code info for extension range options

jhump opened the pull request:
Fixes #3819 

I broke this up into multiple commits to make the changes easier to follow:
1. Compute paths (in the `SourceCodeInfo.Location` meaning of path) to options when building the file descriptor.
2. Use the computed paths from step 1 to construct a map from "path to uninterpreted option" to "path to interpreted option" during the option interpretation step.
3. After options are interpreted, apply the mapping constructed in step 2 to modify the file's source code info.

While testing this (with my own Go test harness that I also used to test my [protoparse](https://godoc.org/github.com/jhump/protoreflect/desc/protoparse) package), I discovered that protoc was also constructing the paths to extension range options incorrectly. The paths that ended up in the source code info were missing the extension range index. So I added a fourth commit so that things looked good:

4. Fix the way paths to extension range options are generated.

I have not yet added a test in this repo. I am a little uncertain of *how* I should test it. FWIW, in my custom Go test harness, I simply process a proto source file that has lots of options (normal and custom options) across various descriptor types (including on extension ranges). My harness can print a human-readable interpretation of the source code info which I visually inspect. I can then commit the results as *golden test files* and have a unit test just process the same source file and verifies that the output matches the "blessed" golden result.

I have looked through some of the existing test code, expecting/hoping to find similar tests (that match outputs against "golden test files") but couldn't quite find any. Since my changes impact both parsing (for the extension range options fix) and descriptors (for interpreting options and re-writing source code info), I think I probably need two tests: (1) a test in `compiler/parser_unittest.cc` that parses source with extension range options and verifies the resulting source code info; and (2) a test in `descriptor_unittest.cc` that builds a descriptor using a proto with uninterpreted options and corresponding source code info and then verifies that the source code info is modified correctly.

If someone can advise, maybe provide pointers to similar kinds of tests, I would greatly appreciate.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/protobuf/4342)
<!-- Reviewable:end -->

jhump commented:
Ping, @xfxyjwf

jhump commented:
Ping @xfxyjwf 
Mind taking a look? Feedback appreciated.

xfxyjwf commented:
@jhump Sorry for delay. I probably won't have time to review this change until late this week though.

jhump commented:
> I probably won't have time to review this change until late this week though.

@xfxyjwf, no problem. I'm just trying to keep it from totally falling off the radar :)

jhump commented:
Ping?

xfxyjwf commented:
Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/google/protobuf/descriptor.cc, line 4158 at r1](https://reviewable.io:443/reviews/google/protobuf/4342#-L7CM0nymBfXMLjbabX0:-L7CM0nymBfXMLjbabX1:bmlxleg) ([raw file](https://github.com/google/protobuf/blob/3174b1e8d432690ea92480712fd77e0df0edc74d/src/google/protobuf/descriptor.cc#L4158)):*
> ```cpp
>     options_to_interpret_.push_back(
>         OptionsToInterpret(name_scope, element_name, element_path_, &orig_options, options));
>   }
> ```

Can we instead use DescriptorT::GetLocationPath() to reconstruct the path?
https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.h#L440

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/protobuf/4342)*
<!-- Sent from Reviewable.io -->

xfxyjwf commented:
I'm using reviewable for this PR as it's more convenient for large changes. Let me know if you have issue viewing the comment.

jhump commented:
*[src/google/protobuf/descriptor.cc, line 4158 at r1](https://reviewable.io:443/reviews/google/protobuf/4342#-L7CM0nymBfXMLjbabX0:-L7Gm_dDkZPyi7RhQ4fy:b18qn8i) ([raw file](https://github.com/google/protobuf/blob/3174b1e8d432690ea92480712fd77e0df0edc74d/src/google/protobuf/descriptor.cc#L4158)):*
<details><summary><i>Previously, xfxyjwf (Feng Xiao) wrote…</i></summary><blockquote>

Can we instead use DescriptorT::GetLocationPath() to reconstruct the path?
https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor.h#L440

</blockquote></details>

I revised the code to do that. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/protobuf/4342)*
<!-- Sent from Reviewable.io -->

xfxyjwf commented:
Thanks. The code overall looks good to me. Could you add some test cases for these new locations in descriptor_unittest.cc?
https://github.com/google/protobuf/blob/master/src/google/protobuf/descriptor_unittest.cc#L6948

You probably need to use FileDescriptor::GetSourceLocation(path, location) with manually constructed paths in the test.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/google/protobuf/descriptor.cc, line 3407 at r2](https://reviewable.io:443/reviews/google/protobuf/4342#-L7QfT8sUGKEb_WE8dCe:-L7QfT8sUGKEb_WE8dCf:bxpmv2b) ([raw file](https://github.com/google/protobuf/blob/11cb912f4cbd054b3a8e95710d919e077bbf1e1b/src/google/protobuf/descriptor.cc#L3407)):*
> ```cpp
>     // must be the mutable copy of the original options message to which
>     // uninterpreted_option_ belongs.
>     bool InterpretSingleOption(Message* options, std::vector<int>& src_path,
> ```

Please add some comments to explain what src_path/options_path are.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/protobuf/4342#-:-L7QktMYRc0KVyIjgSeP:b-79zoc6)*
<!-- Sent from Reviewable.io -->

xfxyjwf commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/google/protobuf/descriptor_unittest.cc, line 7088 at r3](https://reviewable.io:443/reviews/google/protobuf/4342#-L8F-TK4lPcpUiDWUPWr:-L8F-TK4lPcpUiDWUPWs:b-r40wwd) ([raw file](https://github.com/google/protobuf/blob/9427fdd95ec11178f4b236fa5e6e46224d5c6653/src/google/protobuf/descriptor_unittest.cc#L7088)):*
> ```cpp
>                    0};
> 
>     std::vector<int> vpath(path, path+2);
> ```

minor style issue: "+" should be surrounded with spaces.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/protobuf/4342)*
<!-- Sent from Reviewable.io -->

jhump commented:
@xfxyjwf I squashed into one commit and rewrote message.

However, you can actually do this yourself when you merge. In case you didn't know: the button in Github has a drop-down next to it that allows for three options (and you can actually change repo settings to not allow all of them):
1. Create a merge commit
2. Squash and merge
3. Rebase and merge
If you do "squash and merge", it gives you a chance to edit the commit message for the squashed commit that will be made to master. And since this branch is in my own fork, not your repo, you don't have to worry about the state of history on my PR branch.

That way, you don't need contributors to do any sort of history clean-up in their forks before accepting their contributions.

xfxyjwf commented:
@jhump I usually just click the "Merge pull request" button and haven't used the other options before. Thanks for explaining it to me. I'll try that next time.

jhump commented:
@xfxyjwf, I don't see any possible link between those MacOS Ruby build issues and this PR. So hopefully this branch is good to go?

jhump commented:
Whoops. I realized I didn't address your last style nit. And I also find a small bug in my change to `parser.cc` that misattributes the location of options for extension ranges. Another commit incoming soon.

jhump commented:
@xfxyjwf, I think this is ready to go now.

jhump commented:
Thanks for reviewing and merging @xfxyjwf!!
